File: Tinynet_objc/Tinynet.m
//
//  Tinynet.m
//  Tinynet
//
//  Created by wayne on 15/5/26.
//  Copyright (c) 2015年 wayne. All rights reserved.
//

#import "Tinynet.h"

NSString * const kGET   = @"GET";
NSString * const kPOST  = @"POST";

static NSTimeInterval kTimeInterval = 10;

@implementation Tinynet

+ (void)setTimeoutInterval:(NSTimeInterval)timeInterval {
    kTimeInterval = timeInterval;
}

+ (void)requestWithURL:(NSString *)inUrl method:(NSString *)inMethod params:(NSDictionary *)inParams completionHandler:(NetCallBack)inCompletionHandler {
    TinynetManager *manager = [[TinynetManager alloc] initWithURL:inUrl method:inMethod params:inParams completionHandler:inCompletionHandler];
    [manager fire];
}

+ (void)requestWithURL:(NSString *)inUrl method:(NSString *)inMethod completionHandler:(NetCallBack)inCompletionHandler {
    TinynetManager *manager = [[TinynetManager alloc] initWithURL:inUrl method:inMethod params:nil completionHandler:inCompletionHandler];
    [manager fire];
}

+ (void)get:(NSString *)inUrl params:(NSDictionary *)inParams completionHandler:(NetCallBack)inCompletionHandler {
    TinynetManager *manager = [[TinynetManager alloc] initWithURL:inUrl method:kGET params:inParams completionHandler:inCompletionHandler];
    [manager fire];
}

+ (void)get:(NSString *)inUrl completionHandler:(NetCallBack)inCompletionHandler {
    TinynetManager *manager = [[TinynetManager alloc] initWithURL:inUrl method:kGET params:nil completionHandler:inCompletionHandler];
    [manager fire];
}

+ (void)post:(NSString *)inUrl params:(NSDictionary *)inParams completionHandler:(NetCallBack)inCompletionHandler {
    TinynetManager *manager = [[TinynetManager alloc] initWithURL:inUrl method:kPOST params:inParams completionHandler:inCompletionHandler];
    [manager fire];
}

+ (void)post:(NSString *)inUrl completionHandler:(NetCallBack)inCompletionHandler {
    TinynetManager *manager = [[TinynetManager alloc] initWithURL:inUrl method:kPOST params:nil completionHandler:inCompletionHandler];
    [manager fire];
}


@end


@implementation TinynetManager
{
    NSString *_method;
    NSString *_url;
    NSDictionary *_params;
    NetCallBack _completionHandler;
    NSURLSession *_session;
    NSMutableURLRequest *_request;
    NSURLSessionTask *_task;
}

- (id)initWithURL:(NSString *)inUrl
           method:(NSString *)inMethod
           params:(NSDictionary *)inParams
completionHandler:(NetCallBack)inCompletionHandler {
    if (self = [super init]) {
        _url = inUrl;
        _method = inMethod;
        _params = inParams;
        _completionHandler = inCompletionHandler;
        _session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
        _request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:_url] cachePolicy:NSURLRequestReloadIgnoringLocalCacheData timeoutInterval:kTimeInterval];
    }
    
    return self;
}

- (void)fire {
    [self _buildRequest];
    [self _buildBody];
    [self _commitTask];
}


- (void)_buildRequest {
    if ([_method isEqualToString:kGET] && _params.count > 0) {
        _request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:[NSString stringWithFormat:@"%@?%@", _url, [self _buildParams:_params]]] cachePolicy:NSURLRequestReloadIgnoringLocalCacheData timeoutInterval:kTimeInterval];
    }
    
    _request.HTTPMethod = _method;
    
    if (_params.count > 0) {
        [_request addValue:@"application/x-www-form-urlencoded" forHTTPHeaderField:@"Content-Type"];
    }
}

- (void)_buildBody {
    if (![_method isEqualToString:kGET] && _params.count > 0) {
        NSError *error = nil;
        NSData *data = [NSJSONSerialization dataWithJSONObject:_params
                                                       options:NSJSONWritingPrettyPrinted
                                                         error:&error];
        if (!error && data) {
            _request.HTTPBody = data;
        }
    }
}

- (void)_commitTask {
    _task = [_session dataTaskWithRequest:_request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        NSError *er = nil;
        id datax = nil;
        if (!error) {
            datax = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:&er];
        }
        NSBlockOperation *op = [NSBlockOperation blockOperationWithBlock:^{
            _completionHandler(er ? nil : datax, response, error);
        }];
        [[NSOperationQueue mainQueue] addOperation:op];
    }];
    [_task resume];
}

- (NSString *)_buildParams:(NSDictionary *)params {
    NSMutableString *components = [NSMutableString new];
    
    for (NSString *itemKey in [params allKeys]) {
        id itemValue = params[itemKey];
        NSString *tmp = [self _queryComponents:itemKey value:itemValue];
        [components appendString:tmp];
    }
    
    return [components substringFromIndex:1];
}

- (NSString *)_queryComponents:(NSString *)inKey value:(id)inValue {
    NSMutableString *components = [NSMutableString new];
    
    if ([inValue isKindOfClass:[NSArray class]]) {
        for (NSString *itemValue in inValue) {
            NSString *tmp = [self _queryComponents:inKey value:itemValue];
            [components appendString:tmp];
        }
    }
    else if ([inValue isKindOfClass:[NSDictionary class]]) {
        for (NSString *itemKey in [inValue allKeys]) {
            id itemValue = inValue[itemKey];
            NSString *tmp = [self _queryComponents:[NSString stringWithFormat:@"%@[%@]", inKey, itemKey] value:itemValue];
            [components appendString:tmp];
        }
    }
    else {
        [components appendFormat:@"&%@=%@", [inKey escape], [[NSString stringWithFormat:@"%@", inValue] escape]];
    }
    
    return components;
}


@end




@implementation NSString (escape)

- (NSString *)escape{
    CFStringRef legalURLCharactersToBeEscaped = CFSTR(":&=;+!@#$()',*");
    return (__bridge_transfer NSString *)CFURLCreateStringByAddingPercentEscapes(nil, (CFStringRef)self, nil, legalURLCharactersToBeEscaped, kCFStringEncodingUTF8);
}

@end

File: Tinynet/Tinynet.swift
//
//  Tinynet.swift
//
//  Created by wayne on 15/5/21.
//  Copyright (c) 2015年 wayne. All rights reserved.
//

import Foundation

public enum Method : String{
    case GET    = "GET"
    case POST   = "POST"
}


public class Tinynet {
    public static var timeoutInterval: NSTimeInterval = 10;
    
    // MARK: request
    public static func request(method: Method, url: String, params: Dictionary<String, AnyObject>, completionHandler: (data: NSData!, response:NSURLResponse!, error: NSError!) -> Void) {
        let manager = TinynetManager(url: url, method: method, params: params, completionHandler: completionHandler)
        manager.fire()
    }
    
    public static func request(method: Method, url: String, completionHandler: (data: NSData!, response:NSURLResponse!, error: NSError!) -> Void) {
        let manager = TinynetManager(url: url, method: method, params: [String:AnyObject](), completionHandler: completionHandler)
        manager.fire()
    }
    
    // MARK: get
    public static func get(url: String, params:Dictionary<String, AnyObject>, completionHandler: (data: NSData!, response:NSURLResponse!, error: NSError!) -> Void) {
        let manager = TinynetManager(url: url, method: Method.GET, params: params, completionHandler: completionHandler)
        manager.fire()
    }
    
    public static func get(url: String, completionHandler: (data: NSData!, response:NSURLResponse!, error: NSError!) -> Void) {
        let manager = TinynetManager(url: url, method: Method.GET, params: [String:AnyObject](), completionHandler: completionHandler)
        manager.fire()
    }
    
    // MARK: post
    public static func post(url: String, params:Dictionary<String, AnyObject>, completionHandler: (data: NSData!, response:NSURLResponse!, error: NSError!) -> Void) {
        let manager = TinynetManager(url: url, method: Method.POST, params: params, completionHandler: completionHandler)
        manager.fire()
    }
    
    public static func post(url: String, completionHandler: (data: NSData!, response:NSURLResponse!, error: NSError!) -> Void) {
        let manager = TinynetManager(url: url, method: Method.POST, params: [String:AnyObject](), completionHandler: completionHandler)
        manager.fire()
    }
}


class TinynetManager {
    private let method: Method!
    private let params: Dictionary<String, AnyObject>
    private let completionHandler: (data: NSData!, response:NSURLResponse!, error: NSError!) -> Void
    private let session = NSURLSession(configuration: NSURLSessionConfiguration.defaultSessionConfiguration())
    private let url: String!
    private var request: NSMutableURLRequest!
    private var task: NSURLSessionDataTask!
    
    // MARK: init
    required init(url: String, method: Method, params: Dictionary<String, AnyObject>, completionHandler: (data: NSData!, response:NSURLResponse!, error: NSError!) -> Void) {
        self.url = url
        self.method = method
        self.params = params
        self.request = NSMutableURLRequest(URL: NSURL(string: url)!)
        self.completionHandler = completionHandler
    }

    // MARK: private
    private func buildRequest() {
        if self.method == Method.GET && self.params.count > 0 {
            print(self.buildParams(self.params))
            self.request = NSMutableURLRequest(URL: NSURL(string: url + "?" + self.buildParams(self.params))!)
        }

        self.request.HTTPMethod = self.method.rawValue
        self.request.timeoutInterval = Tinynet.timeoutInterval;

        if self.params.count > 0 {
            self.request.addValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        }
    }

    private func bulidBody() {
        if self.method != Method.GET && self.params.count > 0 {
            var error: NSError?
            let data: NSData?
            do {
                data = try NSJSONSerialization.dataWithJSONObject(self.params, options: NSJSONWritingOptions.PrettyPrinted)
            } catch let error1 as NSError {
                error = error1
                data = nil
            }
            if error != nil {
                
            } else {
                self.request.HTTPBody = data
            }
        }
    }
    
    private func commitTask() {
        self.task = session.dataTaskWithRequest(self.request, completionHandler: { (data, response, error) -> Void in
            // 主线程回调
            let opBlock = NSBlockOperation(block: { () -> Void in
                self.completionHandler(data: data, response: response, error: error)
            })
            NSOperationQueue.mainQueue().addOperation(opBlock)
        })
        self.task.resume()
    }
    
    internal func fire() {
        self.buildRequest()
        self.bulidBody()
        self.commitTask()
    }

    // 构建参数
    private func buildParams(params: Dictionary<String, AnyObject>) -> String{
        var components = [(String, String)]();
        let arrkeys = Array(params.keys).sort()
        for key in arrkeys {
            // 获取键值
            let value: AnyObject! = params[key];
            components += self.queryComponents(key, value)
        }
        return (components.map({"\($0)=\($1)"}) as [String]).joinWithSeparator("&");
    }

    private func queryComponents(key: String, _ value: AnyObject) -> [(String, String)] {
        var components = [(String, String)]();
        if let dictionary = value as? [String:AnyObject] {
            for (nestedKey, value) in dictionary {
                components += queryComponents("\(key)[\(nestedKey)]", value)
            }
        } else if let array = value as? [AnyObject] {
            for value in array {
                components += queryComponents("\(key)", value)
            }
        } else {
            components.appendContentsOf([(self.escape(key), self.escape("\(value)"))])
        }

        return components
    }

    private func escape(string: String) -> String {
        let legalURLCharactersToBeEscaped = NSCharacterSet.init(charactersInString: ":&=;+!@#$()',*")
        return string.stringByAddingPercentEncodingWithAllowedCharacters(legalURLCharactersToBeEscaped)!
    }


}
